Add Snowflake mobile model v1 (Close #85)

bill-warner opened the pull request:
Couple of points to call out:

- I have added a new procedure, `commit_table`, which replicates the same procedure in GBQ. This is in place of the `commit_staged` which was previously used in the Snowflake web model. I prefer this as it is a more general procedure that can also be used by the users for custom modules.
- I have also modified `column_check`. The `AUTOMIGRATE` input is now a BOOLEAN rather than a VARCHAR. This made more sense to me and again aligns with the GBQ model. The issue here is that this procedures is used in the web model too, and hasnt been updated in that model. However from what I understand Snowflake can have two procedures with the same name but different input types so in theory if the users is running both models the procedures shouldnt clash. Alternatively we could rename this to `column_check_v2` or something, or update the web model. 
- I removed the duplicates step `07-duplicates` in the base module. Given we never actually fully remove dupes, the insert overwrite procedure seemed overkill here just to count how many rows we removed. Again this aligns with GBQ.
- `06-events-this-run`: I removed the stored procedure from the web model that effectively replicates an `EXCEPT` statement in GBQ. Instead, I have chosen to leave in contexts cols like the page_view context. This simplified the logic and can potentially avoid headache later down the line if new fields are added to these columns that are later required.

adatzer commented on mobile/v1/snowflake/sql-runner/sql/custom/04-session-goals/01-session-goals-staged.sql:
```diff
@@ -0,0 +1,35 @@
+-- 1. Aggregate with a drop and recompute logic
```
Do we add copyright to custom module files?

adatzer commented on mobile/v1/snowflake/sql-runner/sql/custom/04-session-goals/02-session-goals-upsert.sql:
```diff
@@ -0,0 +1,11 @@
+-- 2. Commit table procedure handles committing to prod, including table creation, and creation of new columns if 'automigrate' is set to TRUE
```
add copyright here too?

adatzer commented on mobile/v1/snowflake/sql-runner/sql/custom/04-session-goals/99-session-goals-cleanup.sql:
```diff
@@ -0,0 +1,3 @@
+
```
add copyright here too?

adatzer commented on mobile/v1/snowflake/sql-runner/sql/custom/04-session-goals/99-session-goals-cleanup.sql:
```diff
@@ -0,0 +1,3 @@
+
+
+DROP TABLE IF EXISTS {{.scratch_schema}}.session_goals_staged{{.entropy}};
```
Why do we drop the staged table here? Would it make sense to add an `XX-destroy` for custom module?
Even though there is no module consuming this staged table, maybe dropping `staged` tables is normally part of destroy steps?

bill-warner replied:
Good question. My rational here was that `session_goals_staged` is a drop and recompute table anyway, so dropping seemed a bit cleaner than truncating. The `XX-destroy` step would just be a repeat of this clean up step, hence it was removed.

adatzer commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+  SOURCE_TABLE:   the source table name
+  TARGET_SCHEMA:  the schema of the target table
+  TARGET_TABLE:   the target table name
+  AUTOMIGRATE:    whether target tabled gets altered if needed (only 'TRUE' enables)
```
Since `AUTOMIGRATE` is now a boolean, the `(only 'TRUE' enables)` part of the comment could be removed.

adatzer commented on mobile/v1/snowflake/sql-runner/sql/standard/01-base/01-main/00-setup-base.sql:
```diff
@@ -0,0 +1,271 @@
+
+    se_category                 VARCHAR(1000),
+    se_action                   VARCHAR(1000),
+    se_label                    VARCHAR(1000),
```
Would it make sense to make the length of `se_label` configurable? 
Or somehow support this coming change (https://github.com/snowplow-incubator/snowplow-snowflake-loader/issues/134)?

bill-warner replied:
Sorry I missed this and it is a really good point. Does this mean some users atomic events table will have the column definition `se_label VARCHAR(4096)` already?

adatzer replied:
Yes, it is so. There exist users already with `se_label` as `VARCHAR(4096)`.

adatzer commented on mobile/v1/snowflake/README.md:
```diff
@@ -0,0 +1,240 @@
+
+### Prerequisites
+
+[SQL-runner](https://github.com/snowplow/sql-runner) must be installed, and a dataset of mobile events from either the Snowplow [iOS tracker](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/objective-c-tracker/) or [Android tracker](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/android-tracker/) must be available  ...
```
Maybe add a note that minimum SQL Runner version required is 0.9.3?

bill-warner replied:
Nice good spot

adatzer commented on mobile/v1/snowflake/sql-runner/playbooks/standard/00-setup/99-metadata-complete.yml.tmpl:
```diff
@@ -0,0 +1,20 @@
+  :model_version:      snowflake/mobile/1.0.0
+  :model:              mobile
+  :scratch_schema:     scratch
+  :output_schema:      derived
```
Here too, could the `output_schema` be removed?

bill-warner replied:
I think this is required for the complete step.

adatzer replied:
The [metadata-cleanup](https://github.com/snowplow/data-models/blob/release/mobile/snowflake/1.0.0/mobile/v1/snowflake/sql-runner/sql/standard/00-setup/99-complete/99-metadata-cleanup.sql) only drops a `.scratch_schema` table, that's why the `output` one is not needed for this playbook. (btw i noticed this is something that can be removed from the respective playbook of the snowflake web model as well)

adatzer commented on mobile/v1/snowflake/sql-runner/playbooks/custom/04-session-goals/99-session-goals-complete.yml.tmpl:
```diff
@@ -0,0 +1,20 @@
+  :model_version:           snowflake/mobile/1.0.0
+  :model:                   mobile
+  :scratch_schema:          scratch
+  :output_schema:           derived
```
Could the `model` and `output_schema` variables be removed?

bill-warner replied:
Good spot. Both can be

adatzer commented on mobile/v1/snowflake/sql-runner/playbooks/standard/02-screen-views/01-screen-views-main.yml.tmpl:
```diff
@@ -0,0 +1,44 @@
+      :file: standard/02-screen-views/01-main/03-commit-screen-views.sql
+      :template: true
+
+
```
nit, empty lines

adatzer commented on mobile/v1/snowflake/sql-runner/playbooks/standard/03-optional-modules/01-app-errors/01-app-errors-main.yml.tmpl:
```diff
@@ -0,0 +1,45 @@
+      :file: standard/03-optional-modules/01-app-errors/01-main/03-commit-app-errors.sql
+      :template: true
+
+
```
nit, empty lines

adatzer commented on mobile/v1/snowflake/sql-runner/playbooks/standard/03-optional-modules/01-app-errors/README.md:
```diff
@@ -0,0 +1,63 @@
+
+`:output_schema:`      name of derived schema
+
+`:enabled:`      			 boolean - Toggles the module on/off.  
```
nit, extra space

adatzer commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+      throw ERROR;
+
+  }
+  return "ok. commit_staged succeeded.";
```
just a typo: `commit_table` instead of `commit_staged`

adatzer commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+      snowflake.createStatement({sqlText: cols_check}).execute();
+  }
+
+  /* Prepare delete/insert statements */
```
If `AUTOMIGRATE` is set to `FALSE` at the moment the `column_check` is not called. 
Since `AUTOMIGRATE` is passed to `column_check` as a variable, would it make sense to move the call to `column_check` out of the `if` block, so that it runs even if `AUTOMIGRATE` is set to false?
The benefit could be a more helpful error (coming from `column_check` and possibly pointing to automigration status), since at the moment we will only get errors straight from the execution of delete/inserts.

bill-warner replied:
Fair point. Redesigning the functions

adatzer commented on mobile/v1/snowflake/sql-runner/playbooks/custom/04-session-goals/01-session-goals-main.yml.tmpl:
```diff
@@ -0,0 +1,25 @@
+  :password:
+:variables:
+  :model_version:           snowflake/mobile/1.0.0
+  :model:                   mobile
```
Could the `model` variable be removed?

bill-warner replied:
Good spot.

colmsnowplow commented:
> I have also modified column_check. The AUTOMIGRATE input is now a BOOLEAN rather than a VARCHAR. This made more sense to me and again aligns with the GBQ model. The issue here is that this procedures is used in the web model too, and hasnt been updated in that model. However from what I understand Snowflake can have two procedures with the same name but different input types so in theory if the users is running both models the procedures shouldnt clash. Alternatively we could rename this to column_check_v2 or something, or update the web model.

For sure we should update the web model when we can. Even though it's allowed I don't like having two distinct functions with the same name - especially since normal use of this design pattern isn't really what we're doing with it. 

I think raise an issue and see if we can get the change scheduled for the web model sooner rather than later.

colmsnowplow commented:
> 06-events-this-run: I removed the stored procedure from the web model that effectively replicates an EXCEPT statement in GBQ. Instead, I have chosen to leave in contexts cols like the page_view context. This simplified the logic and can potentially avoid headache later down the line if new fields are added to these columns that are later required.

So does this basically mean a breaking change to the web model when we integrate the base module changes into that?

bill-warner commented:
> > 06-events-this-run: I removed the stored procedure from the web model that effectively replicates an EXCEPT statement in GBQ. Instead, I have chosen to leave in contexts cols like the page_view context. This simplified the logic and can potentially avoid headache later down the line if new fields are added to these columns that are later required.
> 
> So does this basically mean a breaking change to the web model when we integrate the base module changes into that?

Technically not because the new base module will output to `web_events_staged` rather then `events_staged`.

bill-warner commented:
> > I have also modified column_check. The AUTOMIGRATE input is now a BOOLEAN rather than a VARCHAR. This made more sense to me and again aligns with the GBQ model. The issue here is that this procedures is used in the web model too, and hasnt been updated in that model. However from what I understand Snowflake can have two procedures with the same name but different input types so in theory if the users is running both models the procedures shouldnt clash. Alternatively we could rename this to column_check_v2 or something, or update the web model.
> 
> For sure we should update the web model when we can. Even though it's allowed I don't like having two distinct functions with the same name - especially since normal use of this design pattern isn't really what we're doing with it.
> 
> I think raise an issue and see if we can get the change scheduled for the web model sooner rather than later.

Ok would you suggest keeping the function naming as is and just align the web model ASAP? The web model release could be a good opportunity to release the integrated base module too.

adatzer reviewed (approved):
LGTM! Great work @bill-warner !

Indeed `AUTOMIGRATE` as a boolean makes much more sense and it is indeed possible to overload the procedures' names in Snowflake, so there is no clash until it is changed in the web model too.
On duplicates, even though their number could still have some value for users (could it indicate a problem?), i do see the point that it may actually be an overkill.

For reference, all tests i ran have succeeded.

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+  To drop:
+  DROP PROCEDURE {{.output_schema}}.commit_table(VARCHAR,VARCHAR,VARCHAR,VARCHAR,VARCHAR,VARCHAR,BOOLEAN);
+*/
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.commit_table(SOURCE_SCHEMA VARCHAR,
```
We should indeed rename this function - since we have two distinct functions with the same arguments of different types, this isn't a safe way to use overloading.

Snowflake will cast the argument if it's the wrong type, which will cause unexpected behaviour: https://docs.snowflake.com/en/sql-reference/stored-procedures-usage.html#overloading-stored-procedure-names

bill-warner replied:
Ok it is actually the `column_check` procedure that was changed between the web and mobile. I am removing this procedure entirely from this version of the model so I don't think there will be any naming clashes now

colmsnowplow replied:
@bill-warner `commit_table` is still best off need being renamed as it is still divergent from the function of the same name in the web models.

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+  return "ok. commit_staged succeeded.";
+
+  // == Helpers ==
+  function list_cols(sch,tbl) {
```
Nit: Not something that has to change this release, but it would be preferable to define our helper functions independently, if possible. An example of why is that currently, we have no way to independently test that `list_cols()` behaves as expected.

bill-warner replied:
Will look to strip out the helper functions into their own functions. On this one specifically, I am not convinced we need it at all

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+  AS
+  $$
+
+  dmls = DML_STATEMENTS.split(';')
```
Is there a reason to split the statements? I'm not seeing any benefit and it intuitively is better to just execute the SQL we are provided rather than arbitrarily splitting it into separate statements. 

If there are a set of related statements which include the delimiter `;`, then `mk_transaction()` won't be able to run them.

Not a blocker for this release, but if there isn't a reason to do it we should look to remove this in future.

bill-warner replied:
I believe we have to be split as Snowflake does not support running multiple sql statements via `execute()`. We could supply multiple statements in an array but I don't think it is a massive improvement.

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+      var columnAdditions = sourceColumns.filter(notIncludedIn(targetColumns));
+
+      if (sourceColumns.some(notIncludedIn(targetColumns)) === true) {
+          if (!AUTOMIGRATE) {
```
Nit: It's counterintuitive to call automigrate here. `column_check()` intuitively should be an atomic function which performs a column check, nothing else. Automigrate should be a feature of `commit_table()`. 

The fact that we have a conditional on automigrate in commit_table as well supports that design.

Another one that's not a blocker for this release, but if it won't delay us too much would be a cleaner design to manage moving forwards.

bill-warner replied:
Agreed. Going to change the structure here

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+  AS
+  $$
+
+  var delim = '~';
```
Nit: Feels like this should take an array rather than a string with an arbitrary delimiter.

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+              // enforce order
+              for (var i = 0; i < targetColumns.length; i++) {
+                  if (targetColumns[i] !== sourceColumns[i]) {
+                      throw "ERROR: Can only migrate extra columns of the end of source"
```
Nit: Error message could be clearer

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,245 @@
+                  }
+              }
+              add_columns_to(TARGET_SCHEMA, TARGET_TABLE, columnAdditions.join(','));
+              return "ok.Columns added."
```
Nit: Being picky here but 'ok' in the success messages (throughout the procedures) seems colloquial and not necessary.

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/01-base/01-main/00-setup-base.sql:
```diff
@@ -0,0 +1,271 @@
+
+-- Create staged table
+-- Not defining mobile_events_staged as table likely to evolve. Suggest we remove web def below in future releases of SF model.
+{{if eq .model "web"}}
```
Why do we have an explicit table definition for web but not mobile?

bill-warner replied:
Ultimately the staged table is going to be mutated by the `commit_table` function as automigrate = true so it seemed redundant to define the staging table. Personally I think it makes sense to remove the web definition. Thoughts?

colmsnowplow replied:
On one hand, it can be useful to have the base fields defined somewhere for dev accessibility purposes. On the other, you're right that it is indeed redundant to do so. I'm happy with either call, I just think it should be consistent. :)

bill-warner replied:
That is a fair point but I think it is slightly confusing to have this strict definition and then happily mutate it later on. I am going to remove.

colmsnowplow reviewed (approved):
LGTM, great work @bill-warner 

I left some nitpicks which can be deferred to another release. 

Approving on the condition that we address the one important change - renaming the overloaded `commit_table()` procedure. :)

colmsnowplow commented:
> Ok would you suggest keeping the function naming as is and just align the web model ASAP? The web model release could be a good opportunity to release the integrated base module too.

Addressed it in my review, but to leave nothing unanswered - we should both rename it and align the web model. The fact that 
a) a commit_table exists already,
 and that 
b) Snowflake will cast a wrong type instead of failing 
converge to leave us in an unsafe position without renaming it. 

In reality we likely won't hit this problem, since they happen to do analogous things, but this is by coincidence rather than design, so we're better off to remove scope for that problem ever manifesting itself.

bill-warner commented:
Ok I think I have made all the agreed upon changes + refactored the stored procedures. Interested to get your thoughts

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -46,121 +46,174 @@ CREATE OR REPLACE PROCEDURE {{.output_schema}}.mk_transaction(DML_STATEMENTS VAR
+  $$
+
+  var additional_cols_stmt = `
+  WITH target_columns AS (
```
Is this roughly the same query as we use in column_checker? If so, can we possibly just call column checker here, and aggregate its results into the string we need instead?

bill-warner replied:
Yeh I looked into that but procedures in Snowflake typically don't return a result set, rather a single value. UDFs allow for result sets but I think we would have to call the UDF twice, which I think would result in two queries

colmsnowplow replied:
OK - it doesn't need to be changed for this release, but this is a single value:

`{"first_result": [], "second_result": []}`

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -46,121 +46,174 @@ CREATE OR REPLACE PROCEDURE {{.output_schema}}.mk_transaction(DML_STATEMENTS VAR
+  To drop:
+  DROP PROCEDURE {{.output_schema}}.alter_table(VARCHAR,VARCHAR,VARCHAR,VARCHAR);
+*/
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.alter_table(SOURCE_SCHEMA VARCHAR,
```
I like the idea of separating this out into its own atomic unit. Good design choice.

colmsnowplow reviewed (approved):
@bill-warner Looking really nice now - I really like what you've done refactoring the procedures, it's much cleaner now.

I do think we still need to rename `commit_table`, but aside from that, it's releasable in my book.

bill-warner commented:
> @bill-warner Looking really nice now - I really like what you've done refactoring the procedures, it's much cleaner now.
> 
> I do think we still need to rename `commit_table`, but aside from that, it's releasable in my book.

Thanks. I still don't understand why we would need to rename `commit_table`. This function is not used within the web version of the model, it calls a function `commit_staged` to upsert the staged data.

colmsnowplow commented:
> Thanks. I still don't understand why we would need to rename commit_table. This function is not used within the web version of the model, it calls a function commit_staged to upsert the staged data.

Ah, I was mistaken, I see what you mean now. You're correct. `commit_table` doesn't exist anywhere else, so yes it's fine to define it here, we don't require that change.

bill-warner commented:
For clarity:
- I have removed the `mk_transaction()` procedure over security concerns and replaced with `update_manifest()` procedure.
- Refactored `alter_table()` to remove the need to query information schema for a second time. 
- Changed `session_id` in `base_session_id_manifest` table to `VARCHAR(128)` rather than `VARCHAR(36)`

colmsnowplow commented on mobile/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql:
```diff
@@ -0,0 +1,307 @@
+  missing_in_target = res.getColumnValue(2);
+  columns_to_add = res.getColumnValue(3);
+
+  var column_results = [];
```
Nit: No reason to initialise and populate the var separately here.

`var column_results = [missing_in_source, missing_in_target, columns_to_add];`

In fact, I'm not sure you need a var at all (unless that's something to do with Snowflake UDFs requiring it). Can you just do this?

`return [missing_in_source, missing_in_target, columns_to_add];`

bill-warner replied:
Will try this now
